docs/designs/data_transform.md

```diff
@@ -1,4 +1,84 @@
-# Data Tranform Design Doc
+# Data Analysis and Tranformation
+
+## The Problem
+
+We want to add data analysis, in addition to data transformation, into the end-to-end machine learning pipeline generated by SQLFlow.
+
+End-to-end machine learning means a pipeline from raw data to model training and applications. With SQLFlow, we propose to describe complex pipelines with SQL in concisely.  An essential part of the solution is to describe the processing of raw data into model inputs. SQLFlow provides a COLUMN clause for this purpose.
+
+Currently, SQLFlow converts content in the COLUMN clause into the Python source code of data transformation.  The following example would call the Feature Column API [`tf/feature_column/categorical_column_with_hash_bucket`](https://www.tensorflow.org/api_docs/python/tf/feature_column/categorical_column_with_hash_bucket) from TensorFlow to convert a string `address` from the database into a single-element tensor of an integer ID.
+
+```sql
+SELECT * FROM welfare TO TRAIN DNNRegressor COLUMN hash_bucket(address, 1000) LABEL income;
+```
+
+However, data transform often requires parameters. The above example requires the bucket size, 1000. Other examples include
+standardize(age, mean, stdev)
+normalize(size, min, max)
+categorize(city, city_name_list)
+
+The challenge is -- users don't want to specify these parameters; instead, they want SQLFlow does the data statistics and derives these parameters from data automatically.  So, the above examples become
+standardize(age)
+normalize(size)
+categorize(city)
+
+In the terminology of TFX Transform, the word *analysis* refers to data statistics that derives the above parameters, given which, the *transforming* refers to the conversion of raw data into model inputs.  This design document is about making SQLFlow support data analysis, in addition to transformation.
+
+
+## The Syntax Extension
+
+We can ask users to write SQL statements for data analysis.  For example, we might want to assume that the following code snippet trains a model to predict the plate price from its size.
+
+```sql
+min, max = SELECT min(age), max(age) FROM plates;
```

Comment:
We can also write SQL statement to do data transformation. How about add the following case for comparison?
```SQL
SELECT 1.00 * (t1.age - t2.age_min) / t2.age_range
FROM plates t1
JOIN
(
    SELECT
        MIN(age) AS age_min,
        MAX(age) - MIN(age) AS age_range
    FROM plates
) t2
ON 1 = 1
```
This transformation described by SQL can't be used directly in model serving.

Reply:
Does this SQL statement really work? If so, your mastery of SQL the language is fantastic!

Reply:
Yes, it works, I have verified this expression in MySQL.